Allow control of finalization garbage collection with `.spec.deletionPolicy`

erikgb opened the pull request:
The `prune` setting of Kustomizations has a two-fold behavior. When set to `true` FLux will:
1. Ensure stale managed resources are deleted.
2. Garbage collect managed resources when Kustomization is deleted (using a finalizer).

We would like to have 1, but not 2, and this is currently not possible. This PR introduces a new API field allowing control over finalization garbage collection independently of `prune`.

### Some background/motivation for this proposed enhancement:

We are using Flux to provision review environments in a multi-tenant environment (shared cluster), and our tenants are not allowed to create Kubernetes namespaces directly. Instead they have to use a custom namespace-scoped resource as a controller for the namespace (and RBAC). This indirect management of namespaces makes it hard to create Kustomizations inside review namespaces. To overcome this, we use the `targetNamespace` field. This works well when the review namespace is created. But when the review namespace is teared down by our review environment automation, there is a finalization race condition between Flux and our namespace controller. If Flux looses this race, the Kustomization is stuck in non-finalized state (because the namespace is gone) and requires manual intervention from a cluster-admin. This is something we definitely would like to avoid! Since the namespace is controlled and eventually deleted by our namespace controller, no resources are in fact orphaned if we disable FLux garbage collection.

Orphaning resources is generally a bad idea, but is sometimes required. 😉 
~~Once we agree on the naming, I will submit a PR to update the docs in the [website repo](https://github.com/fluxcd/website).~~

Close https://github.com/fluxcd/kustomize-controller/issues/662

stefanprodan commented on api/v1/kustomization_types.go:
```diff
@@ -95,6 +99,13 @@ type KustomizationSpec struct {
 
+	// DeletionPolicy can be used to control garbage collection when this
+	// Kustomization is deleted. Will mirror value of Prune if unspecified.
+	// Valid values are ('MirrorPrune' (default), 'Delete')
```
```suggestion
	// Valid values are ('MirrorPrune' (default), 'Delete', 'Orphan')
```

stefanprodan reviewed (commented):
Please add a new section to the API docs right after https://github.com/fluxcd/kustomize-controller/blob/main/docs/spec/v1/kustomizations.md#prune

stefanprodan commented on internal/controller/kustomization_controller.go:
```diff
@@ -956,10 +956,17 @@ func (r *KustomizationReconciler) prune(ctx context.Context,
 	return false, nil
 }
 
+func finalizerShouldDeleteDependents(obj *kustomizev1.Kustomization) bool {
```
```suggestion
func finalizerShouldDeleteResources(obj *kustomizev1.Kustomization) bool {
```

stefanprodan commented on internal/controller/kustomization_controller.go:
```diff
@@ -956,10 +956,17 @@ func (r *KustomizationReconciler) prune(ctx context.Context,
 }
 
+func finalizerShouldDeleteDependents(obj *kustomizev1.Kustomization) bool {
+	if obj.Spec.DeletionPolicy == nil || *obj.Spec.DeletionPolicy == kustomizev1.DeletionPolicyMirrorPrune {
+		return obj.Spec.Prune
+	}
```
Please add a function to the API package called `GetDeletionPolicy()`, similar to `GetRetryInterval()`, that returns the default value when the field value is set to string empty. Then change the logic here to `if obj.GetDeletionPolicy() == kustomizev1.DeletionPolicyMirrorPrune`.

erikgb replied:
This should be fixed now. I will write docs and add more tests before rebasing and squashing everything into a single commit.

stefanprodan commented on api/v1/kustomization_types.go:
```diff
@@ -95,6 +99,13 @@ type KustomizationSpec struct {
+	// Valid values are ('MirrorPrune' (default), 'Delete')
+	// +kubebuilder:validation:Enum=MirrorPrune;Delete;Orphan
+	// +optional
+	DeletionPolicy *string `json:"deletionPolicy,omitempty"`
```
```suggestion
	DeletionPolicy string `json:"deletionPolicy,omitempty"`
```

No need to set a pointer to a string.

matheuscscp commented on internal/controller/kustomization_deletion_policy_test.go:
```diff
@@ -0,0 +1,220 @@
+	kustomizev1 "github.com/fluxcd/kustomize-controller/api/v1"
+)
+
+func TestKustomizationReconciler_DeletionPolicyDelete(t *testing.T) {
```
Please test also `MirrorPrune` with the different possible values for `prune`.

erikgb replied:
@matheuscscp, will do! Are there any recommendations or established patterns in this project to avoid so much duplicated test code? If not, I am pretty good at copy-modify. 😉

matheuscscp replied:
We always try to do the usual table testing widespread in Go projects. However if the test cases are too different then it's OK to have separate `func TestXXX` tests for them. In this specific case you could for example group together the cases where deletion is expected, and make another group where orphaning is expected.

erikgb replied:
Thanks! Do you have a good example of how you write table-driven tests in Flux projects? Assuming there are multiple ways to do this.

erikgb replied:
Thanks! I have updated the tests now. I have some issues running the tests on my workstation (unrelated to this change). I get the following error:

````
panic: Failed to create Vault instance: could not connect to docker: reached retry deadline: Get "http://127.0.0.1:32769/v1/sys/init": dial tcp 127.0.0.1:32769: connect: connection refused
````

So would it be possible to approve workflows to check the tests (and CI in general)?

I also want to rebase this PR before merge. But I'll wait until it's "approved" (almost ready to merge).

matheuscscp reviewed (commented):
Thanks for this contribution! I think it's a really valuable one.

tenstad commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,46 @@ kustomize.toolkit.fluxcd.io/prune: disabled
+This setup requires use of [`.spec.targetNamespace`](#target-namespace) to get
+anything into a new namespace. When deleting both the Kustomization and the
+namespace controller resource, there is a risk of the namespace being gone
+before Flux initializes the Kustomization finalization.
```
Time of completion is what's relevant?

```suggestion
before Flux completes the Kustomization finalization.
```

tenstad commented on api/v1/kustomization_types.go:
```diff
@@ -95,6 +99,13 @@ type KustomizationSpec struct {
 	Prune bool `json:"prune"`
 
+	// DeletionPolicy can be used to control garbage collection when this
+	// Kustomization is deleted. Will mirror value of Prune if unspecified.
+	// Valid values are ('MirrorPrune' (default), 'Delete', 'Orphan')
```
The mapping could be described in the docs instead, that might be helpful (and sufficient).

erikgb replied:
> I do like the convention of a final Defaults to statement

I agree in general. But the default is a bit "special" in this case.

tenstad replied:
Don't see the harm in writing the default _value_ (not behaviour) when the sentence before it describes the behavior of that value.

I think the connection between the two current statements can be improved. _mirror value of Prune_ relates to _MirrorPrune_ and _unspecified_ relates to _default_, but it is not that explicit, which can lead to confusion.

> Will mirror value of Prune if unspecified.

> Valid values are ('MirrorPrune' (default), 'Delete', 'Orphan')

stefanprodan commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,46 @@ kustomize.toolkit.fluxcd.io/prune: disabled
+
+Valid values include:
+
+- `Delete` - Ensure the managed resources are deleted before the Kustomization
+   is finalized.
+- `Orphan` - Leave the managed resources when the Kustomization is deleted.
```
We need to add MirrorPrune first and explain what it does.

erikgb replied:
Made an attempt. Please take a look.

stefanprodan commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,46 @@ kustomize.toolkit.fluxcd.io/prune: disabled
+   is finalized.
+- `Orphan` - Leave the managed resources when the Kustomization is deleted.
+
+Orphaning resources is generally a bad idea which can lead to confusion.
+But if another garbage collection mechanism is in play, it might make sense
+to orphan resources in the context of Flux.
+
+One example is in multi-tenant clusters where a tenant might not be allowed to
+provision namespaces directly, but have to do it indirectly through a custom
+resource type.
+This setup requires use of [`.spec.targetNamespace`](#target-namespace) to get
+anything into a new namespace. When deleting both the Kustomization and the
+namespace controller resource, there is a risk of the namespace being gone
+before Flux completes the Kustomization finalization.
+To avoid the Kustomization from being stuck in a non-finalized state, it can
+make sense to orphan managed resources:
```
```suggestion
For special cases when the managed resources are removed by other means (e.g. the deletion of the namespace specified with `.spec.targetNamespace`), you can set the deletion policy to `Orphan`:
```

stefanprodan commented on internal/controller/kustomization_deletion_policy_test.go:
```diff
@@ -0,0 +1,255 @@
+/*
+Copyright 2021 The Flux authors
```
```suggestion
Copyright 2024 The Flux authors
```

erikgb commented:
Ok, it seems like we are approaching something here. 😸  I have just rebased and squashed all commits. The PR should be ready to merge if you are satisfied with the proposed changes.

tenstad commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,39 @@ kustomize.toolkit.fluxcd.io/prune: disabled
 
+### Deletion policy
+
+`.spec.deletionPolicy` is an optional field that is allowing control over
```
```suggestion
`.spec.deletionPolicy` is an optional field that allows control over
```

tenstad commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,39 @@ kustomize.toolkit.fluxcd.io/prune: disabled
+
+`.spec.deletionPolicy` is an optional field that is allowing control over
+garbage collection when a Kustomization object is deleted. The default behavior
+is to mirror the value of [`.spec.prune`](#prune).
```
Not totally sure about this one, but it's not the concrete true/false value that is mirrored. Should that be addressed here?

```suggestion
is to mirror the configuration of [`.spec.prune`](#prune).
```

tenstad commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,39 @@ kustomize.toolkit.fluxcd.io/prune: disabled
+- `MirrorPrune` (default) - The managed resources will be deleted if `prune` is
+  `true` and orphaned if `false`.
+- `Delete` - Ensure the managed resources are deleted before the Kustomization
+   is finalized.
+- `Orphan` - Leave the managed resources when the Kustomization is deleted.
```
Keep it simple and non-technical, and equal between the two sentances?
```suggestion
   is deleted.
- `Orphan` - Leave the managed resources when the Kustomization is deleted.
```

tenstad commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,39 @@ kustomize.toolkit.fluxcd.io/prune: disabled
+kind: Kustomization
+metadata:
+  name: app
+  namespace: tenant
```
Not relevant here, might as well remove?
```suggestion
```

matheuscscp replied:
I'm ok with removing it but I'd prefer to keep it, otherwise a new user reading this quickly might think `Kustomization` is cluster-scoped. :shrug:

tenstad commented on api/v1/kustomization_types.go:
```diff
@@ -95,6 +99,14 @@ type KustomizationSpec struct {
 
+	// DeletionPolicy can be used to control garbage collection when this
+	// Kustomization is deleted. Valid values are ('MirrorPrune', 'Delete',
+	// 'Orphan'). 'MirrorPrune' mirrors the Prune field (false = 'Ophan',
+	// true = 'Delete'). Defaults to 'MirrorPrune'.
```
If becoming a bit too technical and type-specific, writing it out in text can also work. Up to you!

```suggestion
	// 'Orphan'). 'MirrorPrune' mirrors the Prune field (orphan if false,
	// delete if true). Defaults to 'MirrorPrune'.
```

tenstad reviewed (commented):
Looking good! Just some minor nitpicks/thouhts regarding text.

matheuscscp commented on docs/spec/v1/kustomizations.md:
```diff
@@ -169,6 +169,39 @@ kustomize.toolkit.fluxcd.io/prune: disabled
+  name: app
+  namespace: tenant
+spec:
+  # ...omitted for brevity
```
```suggestion
```

erikgb replied:
This was copied from another example in the docs. Without it, a reader might wonder if this is a valid Kustomization.

matheuscscp commented on internal/controller/kustomization_deletion_policy_test.go:
```diff
@@ -0,0 +1,255 @@
+				return resultK.Status.LastAppliedRevision == revision
+			}, timeout, time.Second).Should(BeTrue())
+
+			t.Run("creates objects", func(t *testing.T) {
```
I believe we can move all these assertions outside the test case. The test cases should be only the ones in the table on the beginning.

erikgb replied:
Done

matheuscscp commented on internal/controller/kustomization_deletion_policy_test.go:
```diff
@@ -0,0 +1,255 @@
+				g.Expect(k8sClient.Get(context.Background(), types.NamespacedName{Name: id, Namespace: id}, resultConfig)).Should(Succeed())
+			})
+
+			t.Run("deletes managed objects", func(t *testing.T) {
```
Ditto.

matheuscscp commented on internal/controller/kustomization_deletion_policy_test.go:
```diff
@@ -0,0 +1,255 @@
+				return resultK.Status.LastAppliedRevision == revision
+			}, timeout, time.Second).Should(BeTrue())
+
+			t.Run("creates objects", func(t *testing.T) {
```
Ditto.

matheuscscp commented on internal/controller/kustomization_deletion_policy_test.go:
```diff
@@ -0,0 +1,255 @@
+				g.Expect(k8sClient.Get(context.Background(), types.NamespacedName{Name: id, Namespace: id}, resultConfig)).Should(Succeed())
+			})
+
+			t.Run("orphans managed objects", func(t *testing.T) {
```
Ditto.

matheuscscp commented on internal/controller/kustomization_deletion_policy_test.go:
```diff
@@ -0,0 +1,255 @@
+	}
+}
+
+func TestKustomizationReconciler_DeletionPolicyOrphan(t *testing.T) {
```
Looking at these two top-level tests I think the difference between them is not huge. I think we could move all the cases into a single top-level test function and add a boolean to the test case struct to indicate whether deletion is expected or not and have and if-else that tests both cases at the end of the test code. Sorry for misleading you on the opposite direction before!

erikgb replied:
No, problem. I will fix it! The same thought struck me when changing this.

matheuscscp reviewed (commented):
Great stuff, I'm ready to give my approval after all the open discussions are resolved!

erikgb commented:
@matheuscscp All discussions should now be resolved. Can you please approve CI to see if I made any mistakes in the test refactoring?

matheuscscp reviewed (approved):
Great contribution! Thanks @erikgb

stefanprodan reviewed (approved):
LGTM

Thanks @erikgb 🏅
